pkg/tarball/tarball.go

```diff
@@ -83,3 +85,76 @@ func DecodeTarball(reader io.Reader, baseDir string) error {
 
 	return nil
 }
+
+// TarDir tars up an entire directory and outputs the tarball to the specified output path.
+// If useGzip is true, it also gzips the resulting tarball.
+func TarDir(dir, outpath string, useGzip bool) error {
+	// ensure the src actually exists before trying to tar it
+	if _, err := os.Stat(dir); err != nil {
+		return errors.Wrapf(err, "tar unable to stat directory %v", dir)
+	}
+
+	outfile, err := os.Create(outpath)
+	if err != nil {
+		return errors.Wrapf(err, "creating tarball %v", outpath)
+	}
+	defer outfile.Close()
+
+	var tw *tar.Writer
+	if useGzip {
+		gzw := gzip.NewWriter(outfile)
+		defer gzw.Close()
+		tw = tar.NewWriter(gzw)
+	} else {
+		tw = tar.NewWriter(outfile)
+	}
+	defer tw.Close()
+
+	return filepath.Walk(dir, func(file string, fi os.FileInfo, err error) error {
+		// Return on any error.
+		if err != nil {
+			return err
+		}
+
+		// Don't include the archive or dir itself.
+		if filepath.Join(dir, fi.Name()) == outpath ||
+			filepath.Clean(file) == filepath.Clean(dir) {
+			return nil
+		}
+
+		// Dirs and files; fix invalid handle issue?
+		if !fi.Mode().IsRegular() && !fi.Mode().IsDir() {
+			return nil
+		}
+
+		// Create a new dir/file header.
+		header, err := tar.FileInfoHeader(fi, fi.Name())
+		if err != nil {
+			return errors.Wrapf(err, "creating file info header %v", fi.Name())
+		}
+
+		// Update the name to correctly reflect the desired destination when untaring.
+		// 1. Remove directory prefix and leading /
+		// 2: Store with Unix path seperators
+		// 3: Clean
+		header.Name = strings.TrimPrefix(path.Clean(filepath.ToSlash(strings.Replace(file, dir, "", -1))), "/")
+		if err := tw.WriteHeader(header); err != nil {
+			return errors.Wrapf(err, "writing header for tarball %v", header.Name)
+		}
+
+		// Only copy contents of regular files
+		if !fi.Mode().IsRegular() {
```

Comment:
Seems like line 125 already does that check.

Reply:
So the first gate allows directories or regular files so that we write headers. This check prevents trying to `os.Copy` on directories which causes an error.